0n0-CART.Rmd

```diff
@@ -0,0 +1,263 @@
+
+```{r}
+confintWidth = function(data){
+  upperAndLowerBounds = quantile(data, c(.025, .095))
+  width = diff(upperAndLowerBounds)
+  names(width) = NULL
+  return(width)
+}
+confintWidth(simData)
+```
+
+```{r}
+m = 2000
+simulatedWidths = numeric(m)
+for(i in 1:m) simulatedWidths[i] = confintWidth(rgamma(n, 2, 1))
+
+quantile(simulatedWidths, c(.025, .975))
+hist(simulatedWidths)
+```
+```{r}
+b = 2000
+bootstrapWidths = numeric(b)
+for(i in 1:b){
+  resample = sample(simData, n, replace = TRUE)
+  bootstrapWidths[i] = confintWidth(resample)
+} 
+
+quantile(simulatedWidths, c(.025, .975))
+quantile(bootstrapWidths, c(.025, .975))
+```
+
+```{r}
+hist(simulatedWidths, prob = TRUE, col = rgb(1, 0, 0, .2))
+hist(bootstrapWidths, prob = TRUE, col = rgb(0, 0, 1, .2), add = TRUE)
+legend("topright", bty = 'n', fill = c("red", "blue"), legend = c("True Mean Dist", "Bootstrap Dist"))
+```
+We can do this with any function at all!
+
+```{r}
+compareBootstrapToTruth = function(functionToCompare, sampleSize = 100, resamples = 1000, plot = FALSE, ret = FALSE){
+  
+  data = rgamma(sampleSize, 2, 1)
+  trueSamples = numeric(resamples)
+  bootstrapSamples = numeric(resamples)
+  if(class(functionToCompare(data)) != "numeric") stop("Function must return a scalar.")
+  
+  for(i in 1:m){
+    trueSamples[i] = functionToCompare(rgamma(sampleSize, 2, 1))
+    resample = sample(data, sampleSize, replace = TRUE)
+    bootstrapSamples[i] = functionToCompare(resample)
+  } 
+  
+  if(plot){
+    trueHist = hist(trueSamples, plot = FALSE)
+    bootHist = hist(bootstrapSamples, plot = FALSE)
+    hist(trueSamples, prob = TRUE, col = rgb(1, 0, 0, .2), main = "", xlab = "Sample Values", ylim = c(0, max(trueHist$density, bootHist$density)))
+    hist(bootstrapSamples, prob = TRUE, col = rgb(0, 0, 1, .2), add = TRUE)
+    legend("topright", bty = 'n', fill = c("red", "blue"), legend = c("True Dist", "Bootstrap Dist"))
+  }
+  out = list(trueSamples = trueSamples, bootstrapSamples = bootstrapSamples)
+  if(ret) return(out)
+}
+
+compareBootstrapToTruth(mean, plot = TRUE)
+compareBootstrapToTruth(confintWidth, plot = TRUE)
+```
+
+# Bagging
+
+Just as bootstrapping can estimate variability for statistics, it can also do so for predictions. This process is known as Bootstrap AGGrigation, or bagging. It turns out that, when doing bootstrap sampling, about 1/3 of the entries don't make it into the resampled data set. These points are called out-of-bag (oob), and the rest are in-bag. We can use oob data like a free testing set. First a simple example to illustrate, then a non-trivial one.
+
+```{r}
+mpgLm = lm(mpg ~ hp + wt, data = mtcars)
+summary(mpgLm)
+predict(mpgLm, mtcars[1,], interval = "confidence")
+
+b = 2000
```

Comment:
you might want to have this split into 2 separate code chunks?
seems like explaining the code here is 2 separate "thoughts"